src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -245,20 +245,40 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
             }
         }
 
-        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer)
+        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer, uint numBytes)
         {
             Debug.Assert(buffer != null);
             Debug.Assert(buffer.Length > 0);
 
+            numBytes = Math.Min(numBytes, (uint)buffer.Length);
```

Comment:
Do we expect wrong `numBytes` to be actually passed in, or are we just doing this defensively?

Reply:
It is done defensively. it was suggested by @ericstj

Reply:
We should add assert and comment to make it clear.

Reply:
Defensively in case the number of bytes provided by Windows does not match the length in the buffer.

Reply:
We had the assert and removed it to add the check :-) I think it is fine to return back the assert.

Reply:
I brought back the assert and I updated the backport PRs to ensure the CI finishes on time before 4pm.

Reply:
>we could be asserting in the break statements added below

The suggestion makes sense to me. Can I add it those asserts in a separate PR?

Reply:
Fine with me. Also, please add comments like: "This is defensive check in case the number of bytes provided by Windows is wrong. We have not seen this situation in practice.".